lib/eco/plugins/manager.py

```diff
@@ -0,0 +1,130 @@
+import os
+import os.path
+
+from api import Plugin
+
+from PyQt4.QtCore import QObject, QTimer, SIGNAL
+
+class PluginManager(QObject):
+    """This object loads and manages Eco plugins.
+
+    The object constructs the Tools menu in the Eco GUI and connects
+    menu actions to plugins.
+
+    Plugins should be stored in ~/.eco and must be subclasses of
+    eco.plugins.api.Plugin.
+    """
+
+    def __init__(self, main_window, tool_plugin_menu):
+        super(PluginManager, self).__init__(parent=None)
+        self._plugins = dict()
+        self._actions = dict()
+        self._main_window = main_window
+        self._plugin_menu = tool_plugin_menu
+        self._plugin_dir = os.path.join(os.path.expanduser("~"), ".eco")
+        self._current_language = None  # Set by self.set_tms()
+
+        # Load plugins and construct UI.
+        self._load_plugins()
+
+        # Use a single shot timer to run external tools, in case the plugins
+        # take a long time to run (i.e. we do not want to restart the timer
+        # while tools are still running in the background). The default
+        # timeout is quite long, because most of the time the user is typing
+        # the code will be syntactically incorrect, and the tools will not
+        # be able to run.
+        self._default_timeout = 2500
+        self.timer = QTimer()
+        self.timer.timeout.connect(self._run_plugins)
+        self.timer.setSingleShot(True)
+        self.timer.start(self._default_timeout)
+
+        # Number of plugins currently running.
+        self._plugins_running = 0
+
+    def _load_plugins(self):
+        """Load plugins and construct GUI elements.
+        """
+        # Load all plugins that can be found on disk.
+        name_space = dict(locals(), **globals())
+        for root, dirs, files in os.walk(self._plugin_dir):
```

Comment:
Do we want to recursively walk the dir? I'd have thought just the top-level (i.e. `os.listdir`) would be less surprising?